[JS Addons] list addons from addons directory

krmanik opened the pull request:
## Pull Request template

## Purpose / Description
This is UI for addons listing in recycler view from addons directory. The list in view will contains  `details`, `remove` and `toggle switch` button.
Also addons menu added to side bar but it will be hidden till all the PR merged.
The implementation will be used in next PR when addons listed from network for download and install. 

<img src="https://user-images.githubusercontent.com/12841290/164175410-f1629fde-6ed6-4c40-858a-dc3e5f18bdd7.png" height=500px></img>

## Fixes
Fixes #7959 

## Approach
1. `dist` package.json made empty when package.json loaded from local directory. 
2. Popup window class created to show info related the current clicked addons
3. Added a class to list addons from directory, the validated AddonModel added to adapter
4. Each view in list have three buttons
`details button` - show popup with addons info like name, author, version etc.
`remove button` - remove the addons from directory
`toggle switch` - used to enable/disable addons, for enabled addons, the content will be added to reviewer or note editor

## How Has This Been Tested?
Tested on emulator and devices. (Test will be added in next PR)

1. Listing of empty folder 
2. Listing when one addon added manually (currently manually, in next PR download and install will be implemented)
3. Info popup window test
4. Toggle switch test
5. Delete test

https://user-images.githubusercontent.com/12841290/164175561-6990f859-a3f0-4ed4-b61d-f0b2d174c4ee.mp4

## Learning (optional, can help others)

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowserAdapter.kt:
```diff
@@ -0,0 +1,135 @@
+        }
+    }
+
+    private fun deleteAddonDir(addonModel: AddonModel, position: Int) {
```
naming: to me the AddonDir is `/addons/`

krmanik replied:
Added a function name `deleteSelectedAddonPackageDir`

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowserAdapter.kt:
```diff
@@ -0,0 +1,135 @@
+        }
+    }
+
+    private fun deleteAddonDir(addonModel: AddonModel, position: Int) {
```
Can we work out a way to do this with a method that doesn't accept a `position`?

krmanik replied:
Removed `position` from the function

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowserAdapter.kt:
```diff
@@ -0,0 +1,135 @@
+
+        addonModel.updatePrefs(preferences!!, addonModel.addonType, true)
+        addonList.remove(addonModel)
+        notifyDataSetChanged()
+        notifyItemRemoved(position)
```
(uninformed) Do we need both?

krmanik replied:
Only `notifyDataSetChanged` can be used.

krmanik commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowser.kt:
```diff
@@ -0,0 +1,109 @@
+                findViewById<LinearLayout>(R.id.no_addons_found_msg).visibility = View.VISIBLE
+            } else {
+                findViewById<LinearLayout>(R.id.no_addons_found_msg).visibility = View.GONE
+            }
```
Thanks, this is interesting implementation

krmanik commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowser.kt:
```diff
@@ -0,0 +1,109 @@
+                val addonPath = File(addonsDir, file.name)
+                val packageJsonPath = File(addonPath, "package/package.json").path
+                val result: Pair<AddonModel?, List<String>> = getAddonModelFromJson(packageJsonPath)
+                val addonModel = result.first!!
```
If addon not null then added to adapter

krmanik commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowser.kt:
```diff
@@ -0,0 +1,109 @@
+
+            mAddonsListRecyclerView.adapter = AddonsBrowserAdapter(mAddonsList)
+        } catch (e: IOException) {
+            Timber.w(e.localizedMessage)
```
Added toast

krmanik commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowserAdapter.kt:
```diff
@@ -0,0 +1,135 @@
+ */
+class AddonsBrowserAdapter(private var addonList: MutableList<AddonModel>) : RecyclerView.Adapter<AddonsBrowserAdapter.AddonsViewHolder>() {
+    private var preferences: SharedPreferences? = null
+    private var context: Context? = null
```
Using this lots of `!!` removed. Thanks

krmanik commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowserAdapter.kt:
```diff
@@ -0,0 +1,135 @@
+        // enabled/disabled status as boolean value in SharedPreferences
+        val jsAddonKey: String = addonModel.addonType
+        val enabledAddonSet = preferences!!.getStringSet(jsAddonKey, HashSet())
+        for (s in enabledAddonSet!!) {
```
This one liner also good.

mikehardy commented on AnkiDroid/src/main/res/layout/addon_details.xml:
```diff
@@ -0,0 +1,210 @@
+                    android:layout_height="1dp"
+                    android:layout_marginBottom="6dp"
+                    android:layout_marginTop="6dp"
+                    android:background="@color/material_blue_grey_100" />
```
are all these color definitions working correctly with the 4 themes?

krmanik replied:
I have carefully selected this color so it works great on all themes.

mikehardy commented on AnkiDroid/src/main/res/menu/navigation_drawer.xml:
```diff
@@ -16,6 +16,11 @@
+            android:id="@+id/nav_addons"
+            android:icon="@drawable/ic_baseline_extension_24"
+            android:title="@string/addons"
+            android:visible="@bool/anki_show_addon_menu" />
```
that's a nice trick in combo with gradle resValue

mikehardy commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -436,4 +436,22 @@
     <string name="malicious_archive_exceeds_limit">Malicious archive. Size exceeds %1$s or contains more than %2$d files</string>
     <string name="file_extract_exceeds_storage_space">File extract exceeds storage space</string>
+
+    <string name="npm_package_name">NPM Package Name</string>
```
I'd like to avoid putting brand names in our strings, NPM is actually a proprietary package repository and a specific brand. I know it is in common usage, but I would like to just have it be "Package Repository" or (*if* I have captured the spirit of the string correctly) just "Addon Package Name". What do you think?

```suggestion
    <string name="addon_package_name">Addon Package Name</string>
```

krmanik replied:
I also think that we should avoid brand name use `Addon` instead.
Thanks.

mikehardy commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -436,4 +436,22 @@
     <string name="file_extract_exceeds_storage_space">File extract exceeds storage space</string>
+
+    <string name="npm_package_name">NPM Package Name</string>
+    <string name="addon_version">Version</string>
```
Just less ambiguous since we're specifying the JS API Version at same time

```suggestion
    <string name="addon_version">Addon Version</string>
```

mikehardy commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -436,4 +436,22 @@
+    <string name="addon_details">Details</string>
+
+    <string name="addons">Addons</string>
+    <string name="javascript_addons">JavaScript Addons</string>
```
I literally had to go look up the capitalization here. W3Schools and Mozilla both agree, that capitalized `S` in `JavaScript` is correct!

mikehardy commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -436,4 +436,22 @@
+    
+    <string name="addon_enabled">%s addon enabled</string>
+    <string name="addon_disabled">%s addon disabled</string>
+    <string name="confirm_remove_addon">Do you want to remove %s ?</string>
```
Who knows how the naming of the add-on is going to be, encasing it in single-quotes should make it clear it's the name

```suggestion
    <string name="confirm_remove_addon">Do you want to remove '%s'?</string>
```

krmanik replied:
Build is failing when quote added.

mikehardy commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -436,4 +436,22 @@
+    <string name="addon_disabled">%s addon disabled</string>
+    <string name="confirm_remove_addon">Do you want to remove %s ?</string>
+    <string name="failed_to_remove_addon">Failed to remove addon. Try again or remove it manually</string>
+    <string name="no_addons_found">No Addons found, Get addons from options menu</string>
```
All the other messages are just capitalizing like normal sentences without treating 'Addons' as something always capitalized

```suggestion
    <string name="no_addons_found">No addons found, Get addons from options menu</string>
```

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowser.kt:
```diff
@@ -0,0 +1,115 @@
+ * When 'Addons' is clicked, it will open Addon Browser screen for listing all addons from directory.
+ */
+class AddonsBrowser : NavigationDrawerActivity(), SubtitleListener {
+    private lateinit var mAddonsList: MutableList<AddonModel>
+    private lateinit var mAddonsListRecyclerView: RecyclerView
```
I think in Kotlin we are leaving the old "Hungarian Notation" style behind, so these can remove the `m` stuff ? @david-allison that's the expected future state in Kotlin right? Just variable names without scoping-information embedded?

david-allison replied:
Let's remove it

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowser.kt:
```diff
@@ -0,0 +1,115 @@
+        initNavigationDrawer(findViewById(android.R.id.content))
+
+        // Add a home button to the actionbar
+        supportActionBar!!.setHomeButtonEnabled(true)
+        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
+        supportActionBar!!.title = getString(R.string.javascript_addons)
```
I hate to see new Kotlin code with `!!` in it, is there any way to avoid this? "There is no way to avoid this" may be a valid answer, I'm just asking/hoping

krmanik replied:
I have tried but there is no option here. Either it should be wrapped in if/else. I have taken this from another class. So, I need to check more implementation for this.

david-allison replied:
For another issue if you agree @mikehardy  

https://github.com/ankidroid/Anki-Android/blob/5da9561a290122b52047c5c24648f078dc02dca7/AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java#L652-L660

If we change `enableToolbar` to return `supportActionBar!!`, we could:

```kotlin
enableToolbar().apply {
        setHomeButtonEnabled(true)
        setDisplayHomeAsUpEnabled(true)
        title = getString(R.string.javascript_addons)
}
```

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowser.kt:
```diff
@@ -0,0 +1,115 @@
+    /**
+     * List addons from directory, for valid package.json the addons will be added to view
+     */
+    private fun listAddonsFromDir() {
```
Since a dir isn't passed in, and it's taking the directory from the current profile really, seems like a more precise name?

Just bikeshedding here, apologies

```suggestion
    private fun listAddonsFromCurrentProfile() {
```

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowserAdapter.kt:
```diff
@@ -0,0 +1,133 @@
+
+        // enabled/disabled status as boolean value in SharedPreferences
+        val jsAddonKey: String = addonModel.addonType
+        val enabledAddonSet = preferences.getStringSet(jsAddonKey, HashSet())
+        holder.addonActivate.isChecked = enabledAddonSet!!.contains(addonModel.name)
```
Is there a way to put a `? HashSet()` at the end of L60 so that L61 doesn't have a `!!` ? I think it's the preferences API that stupidly doesn't guarantee non-null even if you send in the default to send back ?

krmanik replied:
wrapped in if else.

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowserAdapter.kt:
```diff
@@ -0,0 +1,133 @@
+     *
+     * @param addonModel
+     */
+    private fun deleteSelectedAddonPackageDir(addonModel: AddonModel) {
+        // remove the js addon folder
+        val currentAnkiDroidDirectory = CollectionHelper.getCurrentAnkiDroidDirectory(context)
+        val addonsHomeDir = File(currentAnkiDroidDirectory, "addons")
+        val dir = File(addonsHomeDir, addonModel.name)
+
+        val deleted = BackupManager.removeDir(dir)
+
```
I don't want to over-engineer it, but I feel like having the filesystem code mixed in with the UI code will make it harder to test etc - you have AddonModel, Is there an AddonStorage perhaps, where the "get me the add on directory / get me the list of addons in directory / delete the addon" code could go? Then it would be really easy to have a test just for that stuff

krmanik replied:
I have created `AddonStorage` class. It is more cleaner and better implementation. Thanks

mikehardy reviewed (changes requested):
A bunch of unrelated comments :-), I don't see anything really wrong here - a minor structural question, some Kotlin null-typing stuff and some grammar bikeshedding ... Curious for thoughts on all of them

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowserAdapter.kt:
```diff
@@ -0,0 +1,130 @@
+     *
+     * @param addonModel
+     */
+    private fun deleteSelectedAddonPackageDir(addonModel: AddonModel) {
```
I think delete can go in the new extracted AddOnStorage object now

mikehardy reviewed (commented):
I think it's really close now - can put more into the storage class and use it by reference, and should probably do a second refactor as a follow-on to have code from previous PRs use the storage class as well and centralize all the directory / file handling (from the download/install/unpack PRs)

krmanik commented:
I think this is good idea to handle the addons related file/folder related from a centralized class. I will do refactor in next PR. (Added to TODO)

mikehardy commented:
I'm going to move this to draft just as a strategy to manage the PR review queue, and because I know you are focused on exams right now (as you should be!). Cheers

krmanik commented:
I will update the PR later.

mikehardy commented:
Not stale - dormant because of too much work on my end and prioritizing scoped storage but this will all happen. JS addons are the future and this is the infrastructure...

mikehardy commented:
The JS AddOn series needs to stay open. It is perfect for 2.17 once storage is finally done

krmanik commented:
> The JS AddOn series needs to stay open. It is perfect for 2.17 once storage is finally done

I will start implementing this after scope storage.

mikehardy commented:
Pains me greatly that these aren't merged yet and I feel a lot of responsibility there - not sure how @krmanik feels about it but I would owe you or anyone a debt of gratitude if the work here was finally shepherded in. It's good stuff IMHO and would be a big step forward I think

david-allison commented:
@BrayanDSO I believe you had the last conversation regarding addons in the forums. Pinging for your understanding & thoughts

BrayanDSO commented:
Reviewer add-ons won't happen in the desktop version if they aren't 99,99% secure, and Damien apparently doesn't want it even as a per-notetype opt-in setting with a big warning or just stopped replying because he's busy with something else.

To avoid duplicating the bulletproofing job in the mobile versions, the reviewer should ideally be moved to Svelte first.

I do think that there should be a JS API and addons infrastructure as soon as possible so we can take advantage of the work of other developers. If we wait until all of those desktop issues to be solved (probably at least 3 years from now if I'm optimistic with the current development pace), we will lose a lot of work that could be done in the addons ecosystem in that meantime.

There are a lot of developers, including myself, that may have some time and motivation right now, but not in the future. I could have implemented the whole infrastructure when I created the forums post, but I lost two weeks just in discussion, and stopped getting replied. Now I don't have time for it, nor I want to do it anymore in the future.

So I strongly suggest to continue the JS addons work here in AnkiDroid. To avoid issues in the future, like when Anki implemented a different TTS syntax, I suggest consulting Damien and making him compromise with the API, with whatever is saved in the collection, and with the way that addons are distributed.
That way, things won't break if he wants to distribute things in GitHub with zip files instead of npm with tar.gz, or save the addons list ordered by creation date instead of ID (just an invented example, idk what is planned).

lukstbit commented:
Ok, I'll be implementing the code in this PR for now.

krmanik committed:
Add initial code for application addons

Contains the initial code to fetch and parse addons and also the basic
UI to display the addons.

krmanik commented:
Thanks @lukstbit for continuing this PR. This is good features we can support for AnkiDroid. I still busy at the moment but willing to review if I get chance.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonStorage.kt:
```diff
@@ -0,0 +1,77 @@
+     */
+    fun deleteSelectedAddonPackageDir(addonName: String): Boolean {
+        val dir = getSelectedAddonDir(addonName)
+        return BackupManager.removeDir(dir)
```
use a better named API

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowserActivity.kt:
```diff
@@ -0,0 +1,214 @@
+ */
+// TODO ===== EXTRACT RELATED STRINGS FOR TRANSLATION BEFORE RELEASE =====
+class AddonsBrowserActivity : AnkiActivity() {
+    private val addonsAdapter by lazy {
```
Can we move to a ViewModel here?

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/AddonsBrowserAdapter.kt:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    override fun areContentsTheSame(oldItem: AddonModelUi, newItem: AddonModelUi): Boolean {
+        return oldItem.modelRef.name == newItem.modelRef.name &&
+            oldItem.modelRef.addonType == newItem.modelRef.addonType &&
+            oldItem.modelRef.addonTitle == newItem.modelRef.addonTitle &&
+            oldItem.modelRef.author["name"] == oldItem.modelRef.author["name"] &&
+            oldItem.modelRef.author["url"] == oldItem.modelRef.author["url"] &&
+            oldItem.modelRef.version == newItem.modelRef.version &&
+            oldItem.modelRef.ankidroidJsApi == newItem.modelRef.ankidroidJsApi &&
+            oldItem.modelRef.description == newItem.modelRef.description &&
+            oldItem.modelRef.homepage == newItem.modelRef.homepage &&
+            oldItem.modelRef.license == newItem.modelRef.license &&
+            oldItem.modelRef.dist.tarball == newItem.modelRef.dist.tarball
```
move to the class

david-allison commented on AnkiDroid/src/main/res/layout/activity_addons_browser.xml:
```diff
@@ -0,0 +1,93 @@
+                android:src="@drawable/ic_extension"
+                app:tint="@color/material_blue_grey_100" />
+
+            <TextView
```
FIxedTextVIew

david-allison reviewed (changes requested):
I think we all want this to move forward. Let's do it

Given the nature of this feature, I'd like to see a little documentation of folder structures  on the disk, and how profiles will interact with this feature when they're added.

In Anki Desktop, addons are global across profiles. This change doesn't allow for this.

I'd like to think about deep links to addons, and how/whether we'd want to accomplish this

mikehardy commented:
> Given the nature of this feature, I'd like to see a little documentation of folder structures on the disk, and how profiles will interact with this feature when they're added.
> 
> In Anki Desktop, addons are global across profiles. This change doesn't allow for this.

Interesting. With per-profile add-ons you can have the same add-on for everyone but you can also have "AnkiGamify" and "Ank-2-button" (I'm making those up) just for your toddler's profile as they're learning phonics without it affecting yours.

That is, per-profile is the most powerful style IMHO.

If it was one or the other, I'd say per-profile.
If global was important, then I'd say both, and it is similar to when you install software and it asks "do you want to install this for all users of the system or just yourself?"

Deep links? Perhaps we should just get it in first 😆

krmanik commented:
@david-allison @mikehardy 
Thanks for the review and comments, I will start updating this features in coming few days.

krmanik commented:
Android Studio interface is changed, rebased the code, now add changes to the code.

krmanik commented:
@david-allison Thanks for the changes, I have added it to local codebase. I have started implementing the suggested changes, will take some days to implement the features.
